jetstream/external_config.py

```diff
@@ -39,6 +46,99 @@ def validate(self, experiment: jetstream.experimenter.Experiment) -> None:
         Analysis("no project", "no dataset", conf).validate()
 
 
+def check_all_keys_lowercase(config: MutableMapping[str, Any]) -> Dict[Any, Any]:
+    """
+    function to recursively check that configuration keys are lowercase.
+
+    Returns a dictionary with a dict key path to key which is invalid.
+
+    """
+
+    invalid_keys = dict()
+    for key, value in config.items():
+        if isinstance(value, dict):
+            result = check_all_keys_lowercase(value)
+            if result:
+                invalid_keys[key] = result
+
+        if key.lower() != key:
+            invalid_keys[key] = {key: "INVALID_CONFIG_SETTING"}
+
+    return invalid_keys
+
+
+def validate_config_settings(config_file: Path) -> None:
+    """
+    Loads external config file and runs a number of validation steps on it:
+    - Checks that all config keys/settings are lowercase
+    - Checks for missing core config keys
+    - Checks for unexpected core configuration keys
+    - Checks that all segments defined under experiment have configuration in segments section
+    - Checks if metric with custom config is defined in metrics.weekly or metrics.overall fields
+
+    Returns None, if issues found with the configuration an Exception is raised
+    """
+
+    config = toml.loads(config_file.read_text())
+
+    if config_errors := check_all_keys_lowercase(config):
+        err_msg = (
+            "It appears some config settings/keys"
+            f"contain unexpected uppercase letters: {config_errors}"
+        )
+        raise WrongCaseConfigurationException(err_msg)
+
+    required_core_config_keys = (
+        "experiment",
+        "segments",
+        "metrics",
+    )
+    core_config_keys_specified = config.keys()
+
+    # checks for missing core config keys
+    if missing_config_key := (set(required_core_config_keys) - set(core_config_keys_specified)):
+        # TODO: write a test for this
+        err_msg = f"Config key[s] missing: {missing_config_key}"
+        raise MissingKeyConfigurationException(err_msg)
+
+    # checks for unexpected core configuration keys
+    if unexpected_config_keys := (set(core_config_keys_specified) - set(required_core_config_keys)):
+        # TODO: write a test for this
+        err_msg = f"Unexpected config key[s] found: {unexpected_config_keys}"
+        raise UnexpectedKeyConfigurationException(err_msg)
+
+    # checks that all segments defined under experiment have configuration in segments section
+    expected_segment_configuration = config["experiment"].get("segments", list())
+    segment_config_keys = config["segments"].keys()
```

Comment:
and only do this check if we have segments defined

Reply:
So quick question about this config:
https://github.com/mozilla/jetstream-config/blob/main/keep-in-dock.toml

Currently, I have one test failing (an existing one) in combination with my new logic as this config appears to be specifying segments in the `experiment` section without actually providing any configuration for them. If I understood correctly this is required?

Error:
```
E                   jetstream.errors.SegmentsConfigurationException: It appears some configuration for specified segments is missing: {'new_or_resurrected_v3', 'regular_users_v3'}. Please refer to https://experimenter.info/jetstream/configuration#defining-segments for Segments configuration guide.
```

Reply:
This would mean this config is not correct? I just want to double check before I try to change anything there or more specifically ask the owner of the config to help me fix it.

Reply:
That is correct, however the segments are defined in mozanalysis: https://github.com/mozilla/mozanalysis/blob/a902079aeeeace0dac72bf2da74ebf19bd8f6461/src/mozanalysis/segments/desktop.py#L18-L42

Reply:
The config is correct. Users can specify segments and metrics that are part of mozanalysis.

Reply:
Indeed. There should be a few more other configs that do the same with metrics